src/Ledger/Ratify/Properties.agda

```diff
@@ -8,31 +8,101 @@ module Ledger.Ratify.Properties (txs : _) (open TransactionStructure txs) where
+  RATIFY'-completeness-helper (no _) (no _) _ _ _ _          (RATIFY-Continue _)   = refl
+  RATIFY'-completeness-helper (no _) (yes _) _ _ _ _         (RATIFY-Reject _ _)   = refl
+  RATIFY'-completeness-helper (yes _) _ (no _) (just _) eq _ (RATIFY-Accept _ _ h) = case trans (sym eq) (completeness _ _ _ _ h) of λ where refl → refl
+  RATIFY'-completeness-helper (yes _) _ (no _) nothing _ _   (RATIFY-Continue _)   = refl
+  RATIFY'-completeness-helper (yes _) _ (yes _) _ _ _        (RATIFY-Continue _)   = refl
+
+  RATIFY'-completeness-helper (no ¬acc) _ _ _ _ _ (RATIFY-Accept acc _ _)                 = ⊥-elim (¬acc acc)
+  RATIFY'-completeness-helper _ _ (yes del) _ _ _ (RATIFY-Accept _ ¬del _)                = ⊥-elim (¬del del)
+  RATIFY'-completeness-helper _ _ _ nothing eq _  (RATIFY-Accept _ _ h)                   = ⊥-elim (nothing⇒∀¬STS eq _ h)
+  RATIFY'-completeness-helper (yes acc) _ _ _ _ _ (RATIFY-Reject ¬acc _)                  = ⊥-elim (¬acc acc)
+  RATIFY'-completeness-helper _ (no ¬exp) _ _ _ _ (RATIFY-Reject _ exp)                   = ⊥-elim (¬exp exp)
+  RATIFY'-completeness-helper (yes acc) _ _ _ _ _ (RATIFY-Continue (inj₁ (¬acc , _)))     = ⊥-elim (¬acc acc)
+  RATIFY'-completeness-helper _ (yes exp) _ _ _ _ (RATIFY-Continue (inj₁ (_ , ¬exp)))     = ⊥-elim (¬exp exp)
+  RATIFY'-completeness-helper (no ¬acc) _ _ _ _ _ (RATIFY-Continue (inj₂ (acc , _)))      = ⊥-elim (¬acc acc)
+  RATIFY'-completeness-helper _ _ (no ¬del) _ _ _ (RATIFY-Continue (inj₂ (_ , inj₁ del))) = ⊥-elim (¬del del)
+  RATIFY'-completeness-helper _ _ _ (just _) eq _ (RATIFY-Continue (inj₂ (_ , inj₂ h)))   = ⊥-elim (h _ (≡-just⇔STS .Equivalence.to eq))
+
+RATIFY'-total : ∀ {Γ s sig} → ∃[ s' ] Γ ⊢ s ⇀⦇ sig ,RATIFY'⦈ s'
+RATIFY'-total {Γ} {⟦ es , removed , d ⟧ʳ} {sig} =
+  let open RatifyEnv Γ; st = proj₂ sig; open GovActionState st
+  in RATIFY'-total-helper (accepted? Γ es st) (¿ expired currentEpoch st ¿) (delayed? action prevAction es d)
+                          (compute ⟦ sig .proj₁ , treasury , currentEpoch ⟧ᵉ es action) refl
+
+instance
+  Computational-RATIFY' : Computational _⊢_⇀⦇_,RATIFY'⦈_
+  Computational-RATIFY' .computeProof Γ s a = just RATIFY'-total
+  Computational-RATIFY' .completeness Γ ⟦ es , removed , d ⟧ʳ a s' h =
+    let open RatifyEnv Γ; st = proj₂ a; open GovActionState st in
+    cong just (RATIFY'-completeness-helper (accepted? Γ es st) ¿ expired currentEpoch st ¿ (delayed? action prevAction es d)
+                                           (compute ⟦ a .proj₁ , treasury , currentEpoch ⟧ᵉ es action) refl s' h)
+
+Computational-RATIFY : Computational _⊢_⇀⦇_,RATIFY⦈_
+Computational-RATIFY = it
+
+RATIFY-total : ∀ {Γ s sig} → ∃[ s' ] Γ ⊢ s ⇀⦇ sig ,RATIFY⦈ s'
+RATIFY-total = SS⇒BS-total RATIFY'-total
+
+RATIFY-complete : ∀ {Γ s sig s'} → Γ ⊢ s ⇀⦇ sig ,RATIFY⦈ s' → RATIFY-total {Γ} {s} {sig} .proj₁ ≡ s'
+RATIFY-complete h = computational⇒rightUnique it (RATIFY-total .proj₂) h
```

Comment:
This is quite cumbersome, given the evident repetition of common types and terms, c.f. the general review comment for possible solutions.

Reply:
That branch looks good, I'm merging it in here. Then we can upgrade the abstract to opaque in the 2.6.4 migration.